### ISSUE TITLE
Appsync multi-auth on @function mutation disables iam rule

### ISSUE THREAD

**Issue by grigull** [aws-amplify/amplify-cli#3700]
**Describe the bug**
I have a custom Appsync resolver function that has to be both accessible by a logged-in user that is part of the Admin group, while at the same time by a Lambda function.
I have created a custom Mutation resolver for my Appsync: 
```
type Mutation {
updatePaymentStatus(input: String): Response @function(name: "transactResolver-${env}") @auth (rules: [{ allow: private, provider: iam }, { allow: groups, groups: ["Admin"]}])
}

type Response @aws_iam @aws_cognito_user_pools {
  code: Int
  message: String
  body: String
}
```
Everything works when I just have the single ```@auth (rules: [{ allow: private, provider: iam }])``` rule, but when I add the additional groups auth rule the Lambda function is denied access to the Appsync call, ```GraphQL error: Not Authorized to access updatePaymentStatus on type Mutation at new ApolloError```

Not sure if the two authentications cannot simply share the same @function or I am missing a step, As I assumed that this logic would be the same as when @auth directive is used on @model. 

**Amplify CLI Version**
4.16.1

**Comment by djsmedes** [comment]
You may want to check the Actions granted to your lambda's role in the CloudFormation template. No idea if this is related, but I discovered a bug(?) in the CLI leading to my lambda being unauthorized to access AppSync, and was able to find a workaround: https://github.com/aws-amplify/amplify-cli/issues/4306

**Comment by dysan1257** [comment]
Any update on this issue? I just ran into it as well. The generated request mapping template seems to only check for the group membership.

**Comment by SwaySway** [comment]
I was able to reproduce this and it looks like it the generated userPools resolver code is running even when using IAM auth.

Opened a PR for this fix #5014. A workaround for this would be override the resolver and add the authMode check in the resolver code (example below)
```vtl
## [Start] Determine request authentication mode **
#if( $util.isNullOrEmpty($authMode) && !$util.isNull($ctx.identity) && !$util.isNull($ctx.identity.sub) && !$util.isNull($ctx.identity.issuer) && !$util.isNull($ctx.identity.username) && !$util.isNull($ctx.identity.claims) && !$util.isNull($ctx.identity.sourceIp) && !$util.isNull($ctx.identity.defaultAuthStrategy) )
  #set( $authMode = "userPools" )
#end
## [End] Determine request authentication mode **
#if( $authMode == "userPools" )
.....<ADD_GENERATED_AUTH_CODE_HERE>
#end

.....<REST_OF_RESOLVER>
```

**Comment by SwaySway** [comment]
Closing this issue as the fix for this is included in the latest release.
